ssl/quic/quic_txpim.c

```diff
@@ -0,0 +1,225 @@
+        OPENSSL_free(n);
+    }
+
+    l->head = l->tail = NULL;
+}
+
+void ossl_quic_txpim_free(QUIC_TXPIM *txpim)
+{
+    if (txpim == NULL)
+        return;
+
+    assert(txpim->in_use == 0);
+    free_list(&txpim->free_list);
+    OPENSSL_free(txpim);
+}
+
+static void list_remove(QUIC_TXPIM_PKT_EX_LIST *l, QUIC_TXPIM_PKT_EX *n)
+{
+    if (l->head == n)
+        l->head = n->next;
+    if (l->tail == n)
+        l->tail = n->prev;
+    if (n->prev != NULL)
+        n->prev->next = n->next;
+    if (n->next != NULL)
+        n->next->prev = n->prev;
+    n->prev = n->next = NULL;
+}
+
+static void list_insert_tail(QUIC_TXPIM_PKT_EX_LIST *l, QUIC_TXPIM_PKT_EX *n)
+{
+    n->prev = l->tail;
+    n->next = NULL;
+    l->tail = n;
+    if (n->prev != NULL)
+        n->prev->next = n;
+    if (l->head == NULL)
+        l->head = n;
+}
+
+static QUIC_TXPIM_PKT_EX *txpim_get_free(QUIC_TXPIM *txpim)
+{
+    QUIC_TXPIM_PKT_EX *ex = txpim->free_list.head;
+
+    if (ex != NULL)
+        return ex;
+
+    ex = OPENSSL_zalloc(sizeof(*ex));
+    if (ex == NULL)
+        return NULL;
+
+    list_insert_tail(&txpim->free_list, ex);
+    return ex;
+}
+
+static void txpim_clear(QUIC_TXPIM_PKT_EX *ex)
+{
+    memset(&ex->public.ackm_pkt, 0, sizeof(ex->public.ackm_pkt));
+    ossl_quic_txpim_pkt_clear_chunks(&ex->public);
+    ex->public.retx_head                   = NULL;
+    ex->public.fifd                        = NULL;
+    ex->public.had_handshake_done_frame    = 0;
+    ex->public.had_max_data_frame          = 0;
+    ex->public.had_max_streams_bidi_frame  = 0;
+    ex->public.had_max_streams_uni_frame   = 0;
+    ex->public.had_ack_frame               = 0;
+}
+
+QUIC_TXPIM_PKT *ossl_quic_txpim_pkt_alloc(QUIC_TXPIM *txpim)
+{
+    QUIC_TXPIM_PKT_EX *ex = txpim_get_free(txpim);
+
+    if (ex == NULL)
+        return NULL;
+
+    txpim_clear(ex);
+    list_remove(&txpim->free_list, ex);
+    ++txpim->in_use;
+    return &ex->public;
+}
+
+void ossl_quic_txpim_pkt_release(QUIC_TXPIM *txpim, QUIC_TXPIM_PKT *fpkt)
+{
+    QUIC_TXPIM_PKT_EX *ex = (QUIC_TXPIM_PKT_EX *)fpkt;
+
+    assert(txpim->in_use > 0);
+    --txpim->in_use;
+    list_insert_tail(&txpim->free_list, ex);
+}
+
+void ossl_quic_txpim_pkt_add_cfq_item(QUIC_TXPIM_PKT *fpkt,
+                                      QUIC_CFQ_ITEM *item)
+{
+    item->pkt_next = fpkt->retx_head;
+    item->pkt_prev = NULL;
+    fpkt->retx_head = item;
+}
+
+void ossl_quic_txpim_pkt_clear_chunks(QUIC_TXPIM_PKT *fpkt)
+{
+    QUIC_TXPIM_PKT_EX *ex = (QUIC_TXPIM_PKT_EX *)fpkt;
+
+    ex->num_chunks = 0;
+}
+
+int ossl_quic_txpim_pkt_append_chunk(QUIC_TXPIM_PKT *fpkt,
+                                     const QUIC_TXPIM_CHUNK *chunk)
+{
+    QUIC_TXPIM_PKT_EX *ex = (QUIC_TXPIM_PKT_EX *)fpkt;
+    QUIC_TXPIM_CHUNK *new_chunk;
+    size_t new_alloc_chunks = ex->alloc_chunks;
+
+    if (ex->num_chunks == ex->alloc_chunks) {
+        new_alloc_chunks = (ex->alloc_chunks == 0) ? 4 : ex->alloc_chunks * 2;
```

Comment:
There is research indicating that a scale factor around but below the golden ratio is the ideal for increasing buffer sizes.  Refer _Why Are Vectors Efficient?_ by Andrew Koenig in JOOP 11(5) 1998

Our priority queue and stacks use an expansion of 8 / 5 and are careful to avoid overflows.  I.e.
```c
new_alloc_chunks = safe_muldiv_size_t(ex->alloc_chunks, 8, 5, &err);
```
However, `MAX_ALLOC_CHUNKS` is small enough that there is no problem with just:
```c
new_alloc_chunks = ex->alloc_chunks * 8 / 5;
```